### DISCUSSION TITLE
HowTo send TemplateMessages?

### DISCUSSION THREAD

**Discussion by Diden05** [tulir/whatsmeow#173] [WhatsApp protocol Q&A]
Hi!
Tell me how to send a TemplateMessage with buttons, like here: https://github.com/adiwajshing/Baileys/issues/1424#issuecomment-1080454467

I would be grateful for a code example.

**Comment by Romerito007** [comment]
> Hi! Tell me how to send a TemplateMessage with buttons, like here: [adiwajshing/Baileys#1424 (comment)](https://github.com/adiwajshing/Baileys/issues/1424#issuecomment-1080454467)
> 
> I would be grateful for a code example.

https://github.com/tulir/whatsmeow/issues/96

**Comment by Diden05** [comment]
@Romerito007 I'm not talking about this message, but about this #176

**Comment by hrizal** [comment]
Template only on official API

Pada tanggal Sel, 21 Jun 2022 18.50, Denis Didenko ***@***.***>
menulis:

**Comment by Diden05** [comment]
@hrizal It works in Baileys so it should work here too, or not?

**Comment by hajsf** [comment]
> I would be grateful for a code example

Below is my code:

```go
package responces

import (
	"DigitalAssistance/global"
	"go.mau.fi/whatsmeow"
	waProto "go.mau.fi/whatsmeow/binary/proto"
	"google.golang.org/protobuf/proto"
	"context"
	"fmt"
	"io/ioutil"
)

func Bye(sender string) {

	content, err := ioutil.ReadFile("./documents/Eid.png")
	if err != nil {
		fmt.Println(err)
	}
	resp, err := global.Cli.Upload(context.Background(), content, whatsmeow.MediaImage)
	if err != nil {
		fmt.Println(err)
	}

	hydratedUrlButton := waProto.HydratedTemplateButton_UrlButton{
		UrlButton: &waProto.HydratedURLButton{
			DisplayText: proto.String("زر الرابط"),
			Url:         proto.String("https://www.bankaljazira.com/FFL"),
		},
	}

	hydratedCallButton := waProto.HydratedTemplateButton_CallButton{
		CallButton: &waProto.HydratedCallButton{
			DisplayText: proto.String("Call me"),
			PhoneNumber: proto.String("+1 23456789"),
		},
	}

	hydratedQuickReplyButton := waProto.HydratedTemplateButton_QuickReplyButton{
		QuickReplyButton: &waProto.HydratedQuickReplyButton{
			DisplayText: proto.String("Quick Reply 1"),
			Id:          proto.String("QuickReply1"),
		},
	}
/*
	imageMessage := waProto.ImageMessage{
		Url:      proto.String("https://mmg.whatsapp.net/d/f/AlAF9hFHR7336ScKIqHneB2NpVYsiKWsT05RKTdEUSbl.enc "),
		Mimetype: proto.String("image/jpeg"),
		Caption:  proto.String("الآن بطاقة العملات مجاناً مدى الحياة تقدّم بالحصول عليها من خلال القنوات الالكترونية لبنك الجزيرة أو عبر الفرع"),
	}
*/
	imageMessage := waProto.ImageMessage{
	//	Caption:  proto.String("شكرا و عيدكم مبارك"),
		Mimetype: proto.String("image/png"), // replace this with the actual mime type
		// you can also optionally add other fields like ContextInfo and JpegThumbnail here

		Url:           &resp.URL,
		DirectPath:    &resp.DirectPath,
		MediaKey:      resp.MediaKey,
		FileEncSha256: resp.FileEncSHA256,
		FileSha256:    resp.FileSHA256,
		FileLength:    &resp.FileLength,
	}

	hydratedImage := waProto.HydratedFourRowTemplate_ImageMessage{
		ImageMessage: &imageMessage,
	}

	hydratedFourRowTemplate := waProto.HydratedFourRowTemplate{
	//	HydratedContentText: proto.String("الآن العرض الجديد"),
	//	HydratedFooterText:  proto.String("تطبّق الشروط والأحكام"),
		HydratedButtons: []*waProto.HydratedTemplateButton{
			{
				Index:          proto.Uint32(uint32(1)),
				HydratedButton: &hydratedUrlButton,
			},
			{
				Index:          proto.Uint32(uint32(2)),
				HydratedButton: &hydratedCallButton,
			},
			{
				Index:          proto.Uint32(uint32(3)),
				HydratedButton: &hydratedQuickReplyButton,
			},
		},
		TemplateId: proto.String("id1"),
		Title:      &hydratedImage,
	}

	templateMessage := waProto.TemplateMessage{
		// ContextInfo:      &waProto.ContextInfo{},
		HydratedTemplate: &hydratedFourRowTemplate,
		Format:           nil,
	}

	msg := waProto.Message{
	//	Conversation:    proto.String("Hello"),
		TemplateMessage: &templateMessage,
	}

	jid, ok := global.ParseJID(sender)
	if !ok {
		return
	}
	send, err := global.Cli.SendMessage(jid, "", &msg) // jid = recipient

	if err != nil {
		global.Log.Errorf("Error sending message: %v", err)
	} else {
		global.Log.Infof("Message sent (server timestamp: %s)", send)
	}
}
```

It worked fine with me at WhatsApp desktop app:

<img width="413" alt="image" src="https://user-images.githubusercontent.com/98168280/174895416-31414347-095f-4723-bb21-bc5a3e63af0c.png">


But no idea why it is not appearing at my WhatsApp mobile app, both personal and business accounts!
Try it and tell me if it worked with you.

**Reply by mamur-rezeki** [reply]
My work code :

```go

// Creating template message
this_message := &waProto.Message{
  TemplateMessage: &waProto.TemplateMessage{
    HydratedTemplate: &waProto.HydratedFourRowTemplate{
      Title: &waProto.HydratedFourRowTemplate_HydratedTitleText{
        HydratedTitleText: "The Title",
      },
      TemplateId:          proto.String("template-id"),
      HydratedContentText: proto.String("The Content"),
      HydratedFooterText:  proto.String("The Footer"),
      HydratedButtons: []*waProto.HydratedTemplateButton{

        // This for URL button
        {
          Index: proto.Uint32(1),
          HydratedButton: &waProto.HydratedTemplateButton_UrlButton{
            UrlButton: &waProto.HydratedURLButton{
              DisplayText: proto.String("The Link"),
              Url:         proto.String("https://fb.me/this"),
            },
          },
        },

        // This for call button
        {
          Index: proto.Uint32(2),
          HydratedButton: &waProto.HydratedTemplateButton_CallButton{
            CallButton: &waProto.HydratedCallButton{
              DisplayText: proto.String("Call us"),
              PhoneNumber: proto.String("1234567890"),
            },
          },
        },

        // This is just a quick reply
        {
          Index: proto.Uint32(3),
          HydratedButton: &waProto.HydratedTemplateButton_QuickReplyButton{
            QuickReplyButton: &waProto.HydratedQuickReplyButton{
              DisplayText: proto.String("Quick reply"),
              Id:          proto.String("quick-id"),
            },
          },
        },
      },
    },
  },
}

// Sending message
WaClient.SendMessage(event.Info.Chat, "", this_message)
```

![2022-06-24_03-19](https://user-images.githubusercontent.com/107448645/175381293-0f6b5c08-d98a-4b4d-a72d-c88b4c34f5e3.png)

**Reply by hajsf** [reply]
> My work code

Thanks, worked with me very well.

**Reply by farhan757** [reply]
> Kode kerja saya:
> 
> ```go
> // Creating template message
> this_message := &waProto.Message{
>   TemplateMessage: &waProto.TemplateMessage{
>     HydratedTemplate: &waProto.HydratedFourRowTemplate{
>       Title: &waProto.HydratedFourRowTemplate_HydratedTitleText{
>         HydratedTitleText: "The Title",
>       },
>       TemplateId:          proto.String("template-id"),
>       HydratedContentText: proto.String("The Content"),
>       HydratedFooterText:  proto.String("The Footer"),
>       HydratedButtons: []*waProto.HydratedTemplateButton{
> 
>         // This for URL button
>         {
>           Index: proto.Uint32(1),
>           HydratedButton: &waProto.HydratedTemplateButton_UrlButton{
>             UrlButton: &waProto.HydratedURLButton{
>               DisplayText: proto.String("The Link"),
>               Url:         proto.String("https://fb.me/this"),
>             },
>           },
>         },
> 
>         // This for call button
>         {
>           Index: proto.Uint32(2),
>           HydratedButton: &waProto.HydratedTemplateButton_CallButton{
>             CallButton: &waProto.HydratedCallButton{
>               DisplayText: proto.String("Call us"),
>               PhoneNumber: proto.String("1234567890"),
>             },
>           },
>         },
> 
>         // This is just a quick reply
>         {
>           Index: proto.Uint32(3),
>           HydratedButton: &waProto.HydratedTemplateButton_QuickReplyButton{
>             QuickReplyButton: &waProto.HydratedQuickReplyButton{
>               DisplayText: proto.String("Quick reply"),
>               Id:          proto.String("quick-id"),
>             },
>           },
>         },
>       },
>     },
>   },
> }
> 
> // Sending message
> WaClient.SendMessage(event.Info.Chat, "", this_message)
> ```
> 
> ![2022-06-24_03-19](https://user-images.githubusercontent.com/107448645/175381293-0f6b5c08-d98a-4b4d-a72d-c88b4c34f5e3.png)

why always send to myself number ?

**Reply by hajsf** [reply]
It looks something changed recently, I started getting the same error as reported at: #305

**Reply by hajsf** [reply]
Try wraping it with Viewonce:
```go
			msg := &waProto.Message{
				ViewOnceMessage: &waProto.FutureProofMessage{				
				<--  message here -->
			}}
```
Check #305 resolution

**Reply by farhan757** [reply]
> Try wraping it with Viewonce:
> 
> ```go
> 			msg := &waProto.Message{
> 				ViewOnceMessage: &waProto.FutureProofMessage{				
> 				<--  message here -->
> 			}}
> ```
> 
> Check #305 resolution

thanks bro, success run, but not show if receiver iphone ..

**Comment by codespearhead** [comment]
#307

**Comment by kingmariano** [comment]
Please how do i set a custom message that replies every single message i receive in my dm

**Reply by codespearhead** [reply]
Take a look at [this example](https://github.com/codespearhead/whatsmeow-quickstart).

**Comment by kingmariano** [comment]
How do i create a suggestion interface where by if a user types in something like "/", a list of suggestions would come out from the input.

**Reply by mamur-rezeki** [reply]
That is WhatsApp Business App.

**Comment by fadhildaffa** [comment]
func (r *WhatsappRepo) SendLinkButton(recipient, messageText, url, buttonText string) error {
	// Ensure recipient JID is full format
	jid := types.NewJID(recipient, types.DefaultUserServer)

	// Minimal TemplateID to satisfy WhatsApp
	templateID := "community-template"

	// Construct the message
	msg := &waProto.Message{
		TemplateMessage: &waProto.TemplateMessage{
			HydratedTemplate: &waProto.TemplateMessage_HydratedFourRowTemplate{
				Title: &waProto.TemplateMessage_HydratedFourRowTemplate_HydratedTitleText{
					HydratedTitleText: messageText,
				},
				TemplateID:          proto.String(templateID),
				HydratedContentText: proto.String("Click one of the options below:"),
				HydratedFooterText:  proto.String("Powered by Bot"),
				HydratedButtons: []*waProto.HydratedTemplateButton{
					// URL Button
					{
						Index: proto.Uint32(1),
						HydratedButton: &waProto.HydratedTemplateButton_UrlButton{
							UrlButton: &waProto.HydratedTemplateButton_HydratedURLButton{
								DisplayText: proto.String(buttonText),
								URL:         proto.String(url),
							},
						},
					},
					// Call Button
					{
						Index: proto.Uint32(2),
						HydratedButton: &waProto.HydratedTemplateButton_CallButton{
							CallButton: &waProto.HydratedTemplateButton_HydratedCallButton{
								DisplayText: proto.String("Call Us"),
								PhoneNumber: proto.String("1234567890"),
							},
						},
					},
					// Quick Reply Button
					{
						Index: proto.Uint32(3),
						HydratedButton: &waProto.HydratedTemplateButton_QuickReplyButton{
							QuickReplyButton: &waProto.HydratedTemplateButton_HydratedQuickReplyButton{
								DisplayText: proto.String("Quick Reply"),
								ID:          proto.String("quick-id-1"),
							},
						},
					},
				},
			},
		},
	}

	// Ensure client is connected
	if !r.Client.IsConnected() {
		return fmt.Errorf("WhatsMeow client not connected")
	}

	// Send the message
	_, err := r.Client.SendMessage(context.Background(), jid, msg)
	if err != nil {
		return fmt.Errorf("failed to send message: %w", err)
	}

	return nil
}
hello im using this even tough this code not error just why? the message is not replied like above ? im using personal whatsapp number (not business) is this right ? must wait 24 hours first?

**Reply by ghost** [reply]
huh?

**Reply by fadhildaffa** [reply]
the code was succcessfully executed not any error but the message not appear in recipient

> huh?

**Reply by Hellodebasishsahu** [reply]
Afaik buttons are no longer supported.
